crates/core/src/operations/set_tbl_properties.rs

```diff
@@ -0,0 +1,311 @@
+        {
+            converted_writer_features.insert(WriterFeatures::CheckConstraints);
+        }
+
+        match new_protocol.writer_features {
+            Some(mut features) => {
+                features.extend(converted_writer_features);
+                new_protocol.writer_features = Some(features);
+            }
+            None => new_protocol.writer_features = Some(converted_writer_features),
+        }
+    }
+    if new_protocol.min_reader_version >= 3 {
+        let converted_reader_features = configuration
+            .iter()
+            .filter(|(_, value)| {
+                value.as_ref().map_or(false, |v| {
+                    v.to_ascii_lowercase().parse::<bool>().is_ok_and(|v| v)
+                })
+            })
+            .map(|(key, _)| (*key).clone().into())
+            .filter(|v| !matches!(v, ReaderFeatures::Other(_)))
+            .collect::<HashSet<ReaderFeatures>>();
+        match new_protocol.reader_features {
+            Some(mut features) => {
+                features.extend(converted_reader_features);
+                new_protocol.reader_features = Some(features);
+            }
+            None => new_protocol.reader_features = Some(converted_reader_features),
+        }
+    }
+    new_protocol
+}
+
+impl std::future::IntoFuture for SetTablePropertiesBuilder {
+    type Output = DeltaResult<DeltaTable>;
+
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let mut this = self;
+
+        Box::pin(async move {
+            let mut metadata = this.snapshot.metadata().clone();
+
+            let current_protocol = this.snapshot.protocol();
+            let properties = this.properties;
+
+            let new_protocol = apply_properties_to_protocol(
+                current_protocol,
+                &properties,
+                this.raise_if_not_exists,
+            )?;
+
+            metadata.configuration.extend(
+                properties
+                    .clone()
+                    .into_iter()
+                    .map(|(k, v)| (k, Some(v)))
+                    .collect::<HashMap<String, Option<String>>>(),
+            );
+
+            let final_protocol =
+                convert_properties_to_features(new_protocol, &metadata.configuration);
+
+            let operation = DeltaOperation::SetTableProperties { properties };
+
+            let mut actions = vec![Action::Metadata(metadata)];
+
+            if current_protocol.ne(&final_protocol) {
+                actions.push(Action::Protocol(final_protocol));
+            }
+
+            let commit = CommitBuilder::from(this.commit_properties)
+                .with_actions(actions.clone())
+                .build(
+                    Some(&this.snapshot),
+                    this.log_store.clone(),
+                    operation.clone(),
+                )?
+                .await?;
+
+            this.snapshot.merge(actions, &operation, commit.version)?;
```

Comment:
@ion-elgreco I had to modify this to compile with the latest `main`. I feel uneasy about merging state rather than just reloading from storage but up to you :shrug:

Reply:
Ah yes this piece is outdated now, with the changes I made in the post commit. We are advancing the state there already so the commit already returns the fully updated snapshot :p